modules/atmosphere/rendering/atmospheredeferredcaster.cpp

```diff
@@ -0,0 +1,1595 @@
+
+            program.setUniform("TRANSMITTANCE_W", (int)_transmittance_table_width);
+            program.setUniform("TRANSMITTANCE_H", (int)_transmittance_table_height);
+            program.setUniform("SKY_W", (int)_irradiance_table_width);
+            program.setUniform("SKY_H", (int)_irradiance_table_height);
+            program.setUniform("OTHER_TEXTURES_W", (int)_delta_e_table_width);
+            program.setUniform("OTHER_TEXTURES_H", (int)_delta_e_table_height);
+            program.setUniform("SAMPLES_R", (int)_r_samples);
+            program.setUniform("SAMPLES_MU", (int)_mu_samples);
+            program.setUniform("SAMPLES_MU_S", (int)_mu_s_samples);
+            program.setUniform("SAMPLES_NU", (int)_nu_samples);
+
+            program.setUniform("ModelTransformMatrix", _modelTransform);
+
+            // Object Space
+            glm::dmat4 inverseModelMatrix = glm::inverse(_modelTransform);
+            program.setUniform("dInverseModelTransformMatrix", inverseModelMatrix);
+
+            // The following scale comes from PSC transformations.
+            float fScaleFactor = renderData.camera.scaling().x * pow(10.0, renderData.camera.scaling().y);
+            glm::dmat4 dfScaleCamTransf = glm::scale(glm::dvec3(fScaleFactor));
+            program.setUniform("dInverseScaleTransformMatrix", glm::inverse(dfScaleCamTransf));
+
+            // World to Eye Space in OS
+            program.setUniform("dInverseCamRotTransform", glm::mat4_cast((glm::dquat)renderData.camera.rotationQuaternion()));
+
+            program.setUniform("dInverseSgctEyeToWorldTranform", glm::inverse(renderData.camera.combinedViewMatrix()));
+
+            // Eye Space in OS to Eye Space in SGCT
+            glm::dmat4 dOsEye2SGCTEye = glm::dmat4(renderData.camera.viewMatrix());
+            glm::dmat4 dSgctEye2OSEye = glm::inverse(dOsEye2SGCTEye);
+            program.setUniform("dSgctEyeToOSEyeTranform", dSgctEye2OSEye);
+
+            // Eye Space in SGCT to Projection (Clip) Space in SGCT
+            glm::dmat4 dSgctEye2Clip = glm::dmat4(renderData.camera.projectionMatrix());
+            glm::dmat4 dInverseProjection = glm::inverse(dSgctEye2Clip);
+
+            program.setUniform("dInverseSgctProjectionMatrix", dInverseProjection);
+
+            program.setUniform("dObjpos", glm::dvec4(renderData.position.dvec3(), 1.0));
+            program.setUniform("dCampos", renderData.camera.positionVec3());
+
+            double lt;
+            glm::dvec3 sunPosWorld = SpiceManager::ref().targetPosition("SUN", "SUN", "GALACTIC", {}, _time, lt);
+            glm::dvec4 sunPosObj = glm::dvec4(0.0);
+
+            // Sun following camera position
+            if (_sunFollowingCameraEnabled) {
+                sunPosObj = inverseModelMatrix * glm::dvec4(renderData.camera.positionVec3(), 1.0);
+            }
+            else {
+                if (_renderableClass == RenderablePlanet) {
```

Comment:
That was necessary because the imprecision in the inverse transformations from NDC to object space and how they were handled by each renderable class. Are renderableplanet atmospheres need in the future?